pkg/recorder/recorder.go

```diff
@@ -0,0 +1,78 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package recorder
+
+import (
+	"fmt"
+
+	"github.com/golang/glog"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/kubernetes"
+	typedcorev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/tools/record"
+)
+
+// Provider knows how to generate new event recorders with given name.
+type Provider interface {
+	// NewRecorder returns an EventRecorder with given name.
+	GetEventRecorderFor(name string) (record.EventRecorder, error)
+	// SetClientSet specifies the Clientset of the provider.
+	SetClientSet(cs *kubernetes.Clientset)
+	// SetScheme specifies the Scheme of the provider.
+	SetScheme(scheme *runtime.Scheme)
+}
+
+type provider struct {
+	clientSet        *kubernetes.Clientset
+	scheme           *runtime.Scheme
+	eventBroadcaster record.EventBroadcaster
+}
+
+// NewProvider create a new Provider instance.
+func NewProvider() Provider {
```

Comment:
The `New` pattern is useful when step requires synchronous steps, or setup can fail.  Since we don't have either in this case, I am not sure how much value this provides.

Reply:
We should either take the deps in the New function *or* inject them with Inject functions.  provider doesn't implement either inject.Scheme or inject.Config so it can't automatically get its deps from manager.SetFields(provider).  If we are making the implementation internal, it doesn't really matter which we choose since we can change it without breaking consumers.  `New` is probably the simpler approach for now, so I'd go with that.

e.g. either

```go
// setup clientset and stuff here
New(config *rest.Config, s scheme *runtime.Scheme) { ... }
```
or

```go
// Setup clientset and stuff here
(p *provider) InjectClient() {...}
(p *provider) InjectScheme() {...}
```

Reply:
Yah, I think we do inject for things that cross components, and here we just make it once, so a New function should be enough.